### ISSUE TITLE
KVO-related crash when re-assigning options

### ISSUE THREAD

**Issue by marcinmatczuk** [highcharts/highcharts-ios#162]
Hi,

In our app we're experiencing a KVO-related crash when quickly assigning new `HIChartView`'s `options` property. I think the crash is due to deallocated observer that didn't deregister as observer in its `dealloc` method.
Regarding the use case: we have one chart view and that shows time-series data and additional buttons in view controller's hierarchy to change time window of chart (like in most stocks apps). If user changes time window quickly, at some point we get stacktrace below.

The other question is: what's the recommended minmal setup to effectively reload chart view with new data and new X-axis setup?

```
2019-02-22 12:51:56.975410+0100 Redacted[21490:7830799] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: '<CGDataProvider 0x7fd2aec935f0 [0x109436b68]>: An -observeValueForKeyPath:ofObject:change:context: message was received but not handled.
Key path: isUpdated
Observed object: <HIColor: 0x60000037b640>
Change: {
    kind = 1;
    new = 0;
}
Context: 0x0'
*** First throw call stack:
(
	0   CoreFoundation                      0x00000001091c41bb __exceptionPreprocess + 331
	1   libobjc.A.dylib                     0x00000001081b6735 objc_exception_throw + 48
	2   CoreFoundation                      0x00000001091c4015 +[NSException raise:format:] + 197
	3   Foundation                          0x000000010756ab4c -[NSObject(NSKeyValueObserving) observeValueForKeyPath:ofObject:change:context:] + 161
	4   Foundation                          0x000000010756afeb NSKeyValueNotifyObserver + 332
	5   Foundation                          0x000000010756e696 NSKeyValueDidChange + 489
	6   Foundation                          0x000000010756a5f8 NSKeyValueDidChangeWithPerThreadPendingNotifications + 143
	7   Highcharts                          0x0000000107b68377 -[HIChartsJSONSerializable updateNSObject:newValue:propertyName:] + 263
	8   Highcharts                          0x0000000107c2402e -[HIColor setString:] + 105
	9   Highcharts                          0x0000000107c23a7f -[HIColor initWithString:] + 91
	10  Highcharts                          0x0000000107c23c92 -[HIColor initWithRGBA:green:blue:alpha:] + 81
	11  Highcharts                          0x0000000107c23e9c -[HIColor initWithUIColor:] + 172
	12  Redacted                            0x0000000107009edf $SSo7HIColorC7uiColorABSgSo7UIColorCSg_tcfcTO + 31
	13  Redacted                            0x0000000107008147 $SSo7HIColorC7uiColorABSgSo7UIColorCSg_tcfC + 39
	14  Redacted                            0x0000000107008812 $S8Redacted17ViewControllerC22insertVerticalDashLine33_1B98B3BE33FAD2DC602CE8CCF7C8CD7BLL10timeSeries05chartS012primaryColoryAA04TimeS5ValueV_SaySo8HISeriesCGzSo7UIColorCtF + 210
	15  Redacted                            0x0000000107008689 $S8Redacted17ViewControllerC15addOneSeriesSet33_1B98B3BE33FAD2DC602CE8CCF7C8CD7BLL7options8Trend12primaryColor09secondaryV0ySo9HIOptionsC_AA0bT0VSo7UIColorCAOtFyAA04TimeG5ValueVXEfU0_ + 457
	16  Redacted                            0x000000010700a991 $S8Redacted17ViewControllerC15addOneSeriesSet33_1B98B3BE33FAD2DC602CE8CCF7C8CD7BLL7options8Trend12primaryColor09secondaryV0ySo9HIOptionsC_AA0bT0VSo7UIColorCAOtFyAA04TimeG5ValueVXEfU0_TA + 65
	17  Redacted                            0x0000000107008bb6 $S8Redacted15TimeSeriesValueVs5Error_pIgyzo_ACsAD_pIegnzo_TR + 54
	18  Redacted                            0x000000010700a9db $S8Redacted15TimeSeriesValueVs5Error_pIgyzo_ACsAD_pIegnzo_TRTA + 27
	19  libswiftCore.dylib                  0x0000000109dd4560 $SSTsE7forEachyyy7ElementQzKXEKF + 352
	20  Redacted                            0x0000000107007a9e $S8Redacted17ViewControllerC15addOneSeriesSet33_1B98B3BE33FAD2DC602CE8CCF7C8CD7BLL7options8Trend12primaryColor09secondaryV0ySo9HIOptionsC_AA0bT0VSo7UIColorCAOtF + 7630
	21  Redacted                            0x0000000107005b9b $S8Redacted17ViewControllerC13transformData33_1B98B3BE33FAD2DC602CE8CCF7C8CD7BLL8Trend5after6before12primaryColor09secondaryU0So9HIOptionsCAA0bQ0V_S2dSo7UIColorCAPtF + 139
	22  Redacted                            0x0000000107009bb8 $S8Redacted17ViewControllerC31trendTimeFilterSelectionChangedyyAA05TrendfgC0CFySb_AA0bJ0VSgs5Error_pSgtcfU_ + 712
	23  Redacted                            0x000000010700a247 $S8Redacted17ViewControllerC31trendTimeFilterSelectionChangedyyAA05TrendfgC0CFySb_AA0bJ0VSgs5Error_pSgtcfU_TA + 23
	24  Redacted                            0x0000000107043fad $S8Redacted12ViewModelC19fetchTimeSeriesData3for9timeFrame10completionyAA05PlantB0C_AA0fgfK0CySb_AA0B5TrendVSgs5Error_pSgtctFyAA10SyncResultOcfU_ + 6157
	25  Redacted                            0x00000001070441ec $S8Redacted12ViewModelC19fetchTimeSeriesData3for9timeFrame10completionyAA05PlantB0C_AA0fgfK0CySb_AA0B5TrendVSgs5Error_pSgtctFyAA10SyncResultOcfU_TA + 124
	26  Redacted                            0x00000001070164f3 $S8Redacted10SyncEngineC27performNextOperationInQueue33_EE2DF62CBAC876AE75619A26F2825004LLyyFyAA0B6ResultOcfU_yycfU_ + 435
	27  Redacted                            0x0000000107018279 $S8Redacted10SyncEngineC27performNextOperationInQueue33_EE2DF62CBAC876AE75619A26F2825004LLyyFyAA0B6ResultOcfU_yycfU_TA + 73
	28  Redacted                            0x0000000106fa2bad $SIeg_IeyB_TR + 45
	29  libdispatch.dylib                   0x000000010bd7e595 _dispatch_call_block_and_release + 12
	30  libdispatch.dylib                   0x000000010bd7f602 _dispatch_client_callout + 8
	31  libdispatch.dylib                   0x000000010bd8c99a _dispatch_main_queue_callback_4CF + 1541
	32  CoreFoundation                      0x00000001091293e9 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9
	33  CoreFoundation                      0x0000000109123a76 __CFRunLoopRun + 2342
	34  CoreFoundation                      0x0000000109122e11 CFRunLoopRunSpecific + 625
	35  GraphicsServices                    0x00000001129cc1dd GSEventRunModal + 62
	36  UIKitCore                           0x000000010f09681d UIApplicationMain + 140
	37  Redacted                            0x0000000106ff6e57 main + 71
	38  libdyld.dylib                       0x000000010bdf5575 start + 1
)
```

**Comment by ihnatmoisieiev** [comment]
Hello @marcinmatczuk,
thank you for choosing Highcharts!

We will investigate the issue as soon as possible.

**Comment by ihnatmoisieiev** [comment]
@marcinmatczuk unfortunately, I can't reproduce the issue. I need more info about it. Can you send your initial options and new assigned options after button pressed? What do you want to change in your options? What version of iOS wrapper are you using?

**Comment by marcinmatczuk** [comment]
@ihnatmoisieiev , sorry for late response (vacations).

Initial config using v7.0.3 is very simple:
```
// SNIPPET 1
let options = HIOptions()

options.exporting = HIExporting()
options.exporting.enabled = false

options.legend = HILegend()
options.legend.enabled = false

options.title = HITitle()
options.title.text = ""

options.credits = HICredits()
options.credits.enabled = false

options.tooltip = HITooltip()
options.tooltip.followTouchMove = true
options.tooltip.enabled = true

let style = HICSSObject()
style.fontSize = 16.toCssFontSize()
style.color = "#002a5480" 
style.fontWeight = "normal"
options.noData = HINoData()
options.noData.style = style

options.plotOptions = HIPlotOptions()
options.plotOptions.series = HISeries(name: "")
options.plotOptions.series.marker = HIMarker()
options.plotOptions.series.marker.enabled = false

let xAxis = HIXAxis()
xAxis.tickPixelInterval = 70
xAxis.type = "datetime"
xAxis.lineWidth = 2
xAxis.tickWidth = 2
xAxis.tickColor = gridLineColor
xAxis.gridLineWidth = 2
xAxis.gridLineColor = gridLineColor
xAxis.gridLineDashStyle = "ShortDash"
xAxis.alignTicks = true
xAxis.tickColor = xAxisLineColor
xAxis.tickLength = 24
xAxis.lineColor = xAxisLineColor

let labels = HILabels()
labels.align = "left"
labels.x = 5
labels.y = 14
labels.style = HICSSObject()
labels.style.color = xAxisTextColor
labels.style.fontSize = 12.toCssFontSize()
xAxis.labels = labels

let formats = HIDateTimeLabelFormats()
/* skipping formatting for day month etc */
xAxis.dateTimeLabelFormats = formats

options.xAxis = [xAxis]
```

Once needed data is loaded, above initial options are re-created and next Y-axis is added:
```
// SNIPPET 2
let yAxis = HIYAxis()
yAxis.tickPixelInterval = 50
yAxis.showFirstLabel = false
yAxis.gridLineWidth = gridLineWidth
yAxis.gridLineColor = HIColor.gridLineColor
yAxis.gridLineDashStyle = "ShortDash"
yAxis.labels = HILabels()
yAxis.labels.style = HICSSObject()
yAxis.labels.style.color = primaryColor.toRgbaHexString()
yAxis.labels.align = "right"
yAxis.title = HITitle(text: "")

options.yAxis = [yAxis]

let area = HIArea()
area.yAxis = options.yAxis.endIndex - 1
area.data = trend.timeSeries.map{ [$0.timestampMillis, $0.value as Any] }
area.color = HIColor(uiColor: primaryColor)
area.fillColor = HIColor(uiColor: secondaryColor)
area.lineWidth = chartLineWidth

options.series = [area]

// pseudo code for inserting vertical lines mimicking "gaps" in the plot

for each gap:
    let line = HILine()
    line.color = HIColor(uiColor: primaryColor)
    line.lineWidth = 1
    line.enableMouseTracking = false
    line.data = /* data */
    options.series.append(line)
```

When I change the time filter via button in my view hierarchy, new data is loaded and the 2nd snippet is called for new data - i.e. newly created options are applied with new data.
The crash occurrs when tapping these time filter buttons very quickly after one another.

Did you have a look if Obj-C wrapper objects deregister properly from KVO? Could the call stack in crash report and especially the `HIColor` listed there help investigating the issue?

Thank you in advance for any insights and recommendations how to reload chart with new data more efficiciently.

**Comment by ihnatmoisieiev** [comment]
@marcinmatczuk thank you for more info. I have tried one more time, but, unfortunately, I can't reproduce it.

Yes, I have checked that Obj-C wrapper objects deregister properly from KVO, but we will investigate this problem deeper.

I can propose you another option. Try to use `addYAxis:` and `addSeries:` methods from `HIChartView` object in your second snippet instead of reassign the options, i.e.
```
let yAxis = HIYAxis()
...
self.chartView.add(yAxis)

let area = HIArea()
...
self.chartView.add(area)

for each gap:
    let line = HILine()
    ...
    self.chartView.add(line)
```
